src/api/swarm_api.go

```diff
@@ -0,0 +1,137 @@
+			ForcedHostIPv4:       params.NewAgentParams.DryForcedHostIPV4,
+			ForcedMacAddress:     params.NewAgentParams.DryForcedMacAddress.String(),
+			ForcedHostname:       params.NewAgentParams.DryForcedHostname,
+			FakeRebootMarkerPath: params.NewAgentParams.DryFakeRebootMarkerPath,
+		},
+		ConnectivityConfig: agentConfig.ConnectivityConfig{
+			TargetURL:          params.NewAgentParams.ServiceURL,
+			InfraEnvID:         params.NewAgentParams.InfraEnvID.String(),
+			AgentVersion:       params.NewAgentParams.AgentVersion,
+			PullSecretToken:    params.NewAgentParams.PullSecret,
+			InsecureConnection: true,
+			CACertificatePath:  params.NewAgentParams.Cacert,
+		},
+		LoggingConfig: agentConfig.LoggingConfig{
+			JournalLogging: true,
+		},
+		IntervalSecs: 60,
+	}
+	subprocessCfg := &agentConfig.SubprocessConfig{
+		LoggingConfig: agentConfig.LoggingConfig{
+			TextLogging:    false,
+			JournalLogging: true,
+		},
+		DryRunConfig: agentCfg.DryRunConfig,
+	}
+	installerCfg := &installerConfig.Config{
+		DryRunConfig: installerConfig.DryRunConfig{
+			DryRunEnabled:          true,
+			FakeRebootMarkerPath:   params.NewAgentParams.DryFakeRebootMarkerPath,
+			ForcedHostID:           params.NewAgentParams.DryForcedHostID.String(),
+			DryRunClusterHostsPath: params.NewAgentParams.DryClusterHostsPath,
+		},
+		InfraEnvID:           params.NewAgentParams.InfraEnvID.String(),
+		URL:                  params.NewAgentParams.ServiceURL,
+		AgentImage:           params.NewAgentParams.AgentVersion,
+		PullSecretToken:      params.NewAgentParams.PullSecret,
+		SkipCertVerification: true,
+		CACertPath:           params.NewAgentParams.Cacert,
+	}
+	logSenderCfg := &agentConfig.LogsSenderConfig{
+		AgentConfig:     *agentCfg,
+		InfraEnvID:      params.NewAgentParams.InfraEnvID.String(),
+		TargetURL:       params.NewAgentParams.ServiceURL,
+		PullSecretToken: params.NewAgentParams.PullSecret,
+	}
+	log := util.NewJournalLogger("combined-agent", params.NewAgentParams.DryForcedHostID.String())
+	if err := installerConfig.DryParseClusterHosts(params.NewAgentParams.DryClusterHostsPath, &installerCfg.ParsedClusterHosts); err != nil {
+		log.WithError(err).Error("DryParseClusterHosts")
+		return swarm.NewCreateNewAgentBadRequest().WithPayload(newError(http.StatusBadRequest, err))
+	}
+	newAgent, err := agents.NewAgent(id)
+	if err != nil {
+		log.WithError(err).Warnf("Cannot allocate agent %d id %s", id, params.NewAgentParams.DryForcedHostID.String())
+		return swarm.NewDeleteAgentInternalServerError().WithPayload(newError(http.StatusInternalServerError, err))
+	}
+	go func() {
+		defer agents.SetTerminated(id)
+
+		factory := factories.NewNextStepRunnerFactory(subprocessCfg, installerCfg, logSenderCfg, log)
+		agent.RunAgent(agentCfg, factory, log)
+	}()
+	return swarm.NewCreateNewAgentCreated().WithPayload(newAgent)
+}
+
+func (s *SwarmAPI) DeleteAgent(ctx context.Context, params swarm.DeleteAgentParams) middleware.Responder {
```

Comment:
Shouldn't this stop the goroutine?

Reply:
Agent has many goroutines not just one.  How can we stop them all?

The caller should call this API once the agent has status Terminated

Reply:
Idiomatically you would do this with a WithCancel context passed around everywhere, not sure how much the agent is prepared for this at this point.